File: 김하영/programmers/디스크_컨트롤러.js
// 프로그래머스 디스크컨트롤러
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";

const input = fs.readFileSync(filePath).toString().trim();
const jobs = JSON.parse(input);

class MinHeap {
  constructor() {
    this.heap = [null];
  }

  size() {
    return this.heap.length - 1;
  }

  getMin() {
    return this.heap[1] ? this.heap[1] : null;
  }

  swap(a, b) {
    [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];
  }

  heappush(value) {
    this.heap.push(value);
    let curIdx = this.heap.length - 1;
    let parIdx = (curIdx / 2) >> 0;

    while (curIdx > 1 && this.heap[parIdx] > this.heap[curIdx]) {
      this.swap(parIdx, curIdx);
      curIdx = parIdx;
      parIdx = (curIdx / 2) >> 0;
    }
  }

  heappop() {
    const min = this.heap[1];
    if (this.heap.length <= 2) this.heap = [null];
    else this.heap[1] = this.heap.pop();

    let curIdx = 1;
    let leftIdx = curIdx * 2;
    let rightIdx = curIdx * 2 + 1;

    if (!this.heap[leftIdx]) return min;
    if (!this.heap[rightIdx]) {
      if (this.heap[leftIdx] < this.heap[curIdx]) {
        this.swap(leftIdx, curIdx);
      }
      return min;
    }

    while (
      this.heap[leftIdx] < this.heap[curIdx] ||
      this.heap[rightIdx] < this.heap[curIdx]
    ) {
      const minIdx =
        this.heap[leftIdx] > this.heap[rightIdx] ? rightIdx : leftIdx;
      this.swap(minIdx, curIdx);
      curIdx = minIdx;
      leftIdx = curIdx * 2;
      rightIdx = curIdx * 2 + 1;
    }

    return min;
  }
}

function solution(jobs) {
  const minHeap = new MinHeap();
  jobs.sort((a, b) => a[0] - b[0]);
  let time = 0;
  let answer = 0;
  let idx = 0;

  while (idx < jobs.length || minHeap.size() > 0) {
    if (idx < jobs.length && jobs[idx][0] <= time) {
      minHeap.heappush(jobs[idx++]);
      continue;
    }

    if (minHeap.size() > 0) {
      const [start, duration] = minHeap.heappop();
      time += duration;
      answer += time - start;
    } else {
      time = jobs[idx][0];
    }
  }

  return (answer / jobs.length) >> 0;
}

console.log(solution(jobs));

File: 김유진/30min_Programmers/힙/디스크 컨트롤러.py
import heapq
def solution(jobs):
    answer = 0
    return answer

'''
혜선님 설명
[디스크 컨트롤러]
요청 큐 = deq
대기 큐 = heap

현재 시간 = 0


요청이 왔을 때
- 요청이 온 시간보다 현재 시간이 같거나 작을때
    - 대기 큐에 작업이 없다면 (비어있다면)
        - 요청을 수행 (요청 시간 + 수행 시간 = 현재 시간)
    - 대기 큐에 작업이 있다면
        - 대기 큐에서 수행 시간이 제일 적은 요청을 꺼내서 수행 (현재 시간 + 수행시간 = 현재 시간)
        - 제일 적은 요청을 수행 했는데도 현재 시간이 먼저 온 요청 시간보다 작거나 같다면
            - 요청을 요청 큐에 다시 집어 넣음
        - 제일 적은 요청을 수행했을 때 현재 시간이 먼저 온 요청 시간보다 큰 경우
            - 대기 큐에 현재 요청을 집어 넣음
- 요청이 온 시간보다 현재 시간이 클때
    - 대기 큐에 집어 넣음

위 과정을 요청 큐가 빌 때까지 계속 진행
이후 대기 큐를 돌며 시간 계산 및 평균 시간 계산
'''
